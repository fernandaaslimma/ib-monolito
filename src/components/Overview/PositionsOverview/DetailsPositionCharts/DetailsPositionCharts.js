import React, { Fragment } from "react";

import {
  WrapperChart,
  WrapperSubChart,
  WrapperCard,
  ChartInfo,
  ChartTitle,
  WrapperIcon,
  WrapperSubCard,
  Title,
  ExpandCollapseArrow,
  WrapperButton
} from "./styles";
import SimpleChartPie from "../../../common/SimpleChartPie";
import Card from "../../../common/Card";
import getColorForAsset from "../../../../utils/getColorForAsset";
import Icon from "../../../common/Icon";
import HideableValue from "../../../common/HideableValue";
import { BRL_CURRENCY } from "../../../../utils/constants";
import { redirect } from "../../../../utils/redirect";

import { gray90, gray300 } from "../../../../styles/settings";
import ShimmerLoading from "../../../common/ShimmerLoading";

import { Button } from "react-bocombbm-components";
import { translate } from "../../../../utils/i18n";
import { formatDateToLocale } from "../../../../utils/formatDate";

function DetailsPositionCharts({
  onCollapse,
  assets,
  colorForAsset,
  loading,
  hideValues,
  dataTest,
  setStateAsset,
  checkStateAsset,
  activeChart,
  findAssettype,
  findLoading
}) {
  const renderLoadingMode = () => {
    return (
      <Fragment>
        {[0, 1, 2].map((_, index) => (
          <WrapperChart
            key={index}
            data-test={`${dataTest}_loading_${index}`}
            loading={loading}
          >
            <Card>
              <WrapperCard>
                <ShimmerLoading circle width={48} height={48} />

                <ChartInfo>
                  <ChartTitle>
                    <ShimmerLoading width={80} height={14} />
                  </ChartTitle>
                  <ShimmerLoading width={80} height={12} darker />
                </ChartInfo>
              </WrapperCard>
            </Card>
          </WrapperChart>
        ))}
      </Fragment>
    );
  };

  const buildSubCard = (type, name) => {
    return findAssettype(name) &&
      checkStateAsset(type) &&
      !findLoading(name) ? (
      findAssettype(name).map((asset, index) => (
        <WrapperSubChart key={index} data-test={`${dataTest}_subCard_${index}`}>
          <Card styles={"background: #F9FAFB; box-shadow: none;"}>
            <WrapperCard>
              <SimpleChartPie
                dataTest={`chart_${index}`}
                percentage={asset.portfolioShare ? asset.portfolioShare : 0}
                background={gray90}
                color={getColorForAsset(1)}
                width={48}
              />
              <ChartInfo>
                <ChartTitle
                  data-test="productName_subCard"
                  ellipsis={true}
                  widthTagEllipsis="100%"
                >
                  {asset.name}
                </ChartTitle>
                <HideableValue
                  hide={hideValues}
                  currency={BRL_CURRENCY}
                  value={asset.grossBalance}
                  currencyColor={gray300}
                  currencySize={12}
                />
              </ChartInfo>
            </WrapperCard>
          </Card>
        </WrapperSubChart>
      ))
    ) : (
      <WrapperChart data-test={`${dataTest}_subCard_loading`}>
        <Card styles={"background: #F9FAFB; box-shadow: none;"}>
          <WrapperCard>
            <ShimmerLoading circle width={48} height={48} />
            <ChartInfo>
              <ChartTitle data-test="productName_subCard">
                <ShimmerLoading width={80} height={14} />
              </ChartTitle>
              <ShimmerLoading width={80} height={12} darker />
            </ChartInfo>
          </WrapperCard>
        </Card>
      </WrapperChart>
    );
  };

  return loading ? (
    renderLoadingMode()
  ) : (
    <Fragment>
      {assets.map((position, index) => (
        <WrapperChart key={index} data-test={`${dataTest}_${index}`}>
          <Card>
            <WrapperCard>
              {loading ? (
                <ShimmerLoading circle width={48} height={48} />
              ) : (
                <SimpleChartPie
                  dataTest={`chart_${index}`}
                  percentage={
                    position.portfolioShare ? position.portfolioShare : 0
                  }
                  background={gray90}
                  color={getColorForAsset(
                    colorForAsset ? colorForAsset : index
                  )}
                  width={48}
                />
              )}

              <ChartInfo>
                <ChartTitle data-test="productName">
                  {loading ? (
                    <ShimmerLoading width={30} height={14} />
                  ) : (
                    position.name
                  )}
                </ChartTitle>
                {loading ? (
                  <ShimmerLoading width={30} height={12} darker />
                ) : (
                  <HideableValue
                    hide={hideValues}
                    currency={BRL_CURRENCY}
                    value={position.grossBalance}
                    currencyColor={gray300}
                    currencySize={12}
                  />
                )}
              </ChartInfo>
              {activeChart === 0 && (
                <WrapperIcon
                  onClick={() => {
                    setStateAsset(
                      activeChart === 0
                        ? position.assetType
                        : position.assetClass
                    );
                    onCollapse(
                      index,
                      activeChart === 0
                        ? position.assetType
                        : position.assetClass,
                      position.name
                    );
                  }}
                >
                  <ExpandCollapseArrow
                    collapse={checkStateAsset(
                      activeChart === 0
                        ? position.assetType
                        : position.assetClass
                    )}
                  >
                    <Icon
                      data-test={`arrowIcon_${index}`}
                      type="Arrow"
                      width="16"
                      height="16"
                    />
                  </ExpandCollapseArrow>
                </WrapperIcon>
              )}
            </WrapperCard>
            {checkStateAsset(
              activeChart === 0 ? position.assetType : position.assetClass
            ) && (
              <WrapperSubCard>
                <Title>{`${translate(
                  "OVERVIEW_CONSOLIDATED_DATE"
                )}${formatDateToLocale(position.date)}`}</Title>
                {buildSubCard(
                  activeChart === 0 ? position.assetType : position.assetClass,
                  position.name
                )}
                {activeChart === 0 && (
                  <WrapperButton data-test="showButtonDetailed">
                    <Button
                      onClick={() => redirect(position.route)}
                      disabled={position.route ? false : true}
                      spacing={{
                        top: "s",
                        bottom: "s",
                        right: "s",
                        left: "s"
                      }}
                      type="outline"
                    >
                      {translate("OVERVIEW_REDIRECT_BUTTON")}
                    </Button>
                  </WrapperButton>
                )}
              </WrapperSubCard>
            )}
          </Card>
        </WrapperChart>
      ))}
    </Fragment>
  );
}

export default DetailsPositionCharts;
