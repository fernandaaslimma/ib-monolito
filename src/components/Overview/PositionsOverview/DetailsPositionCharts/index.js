import DetailsPositionCharts from "./DetailsPositionCharts";

export default DetailsPositionCharts;
