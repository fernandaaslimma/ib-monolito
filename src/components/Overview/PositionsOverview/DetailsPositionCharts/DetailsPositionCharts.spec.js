import React from "react";
import { render } from "enzyme";

import DetailsPositionCharts from "./DetailsPositionCharts";

jest.mock("../../../../utils/redirect");

const props = {
  assets: [
    {
      name: "Conta Corrente",
      assetType: "CashAccount",
      assetTypeLabel: "Conta Corrente",
      date: "2018-05-04",
      netBalance: 631409.01,
      grossBalance: 631409.01,
      incomeTaxBalance: 0.71,
      iofBalance: 0,
      grossResultBalance: 631409.01,
      portfolioShare: 26.75,
      route: "mock"
    },
    {
      name: "Conta Corrente",
      assetType: "CashAccount",
      assetTypeLabel: "Conta Corrente",
      date: "2018-05-04",
      netBalance: 631409.01,
      grossBalance: 631409.01,
      incomeTaxBalance: 0.71,
      iofBalance: 0,
      grossResultBalance: 631409.01,
      route: "mock"
    }
  ],
  colorForAsset: 0,
  loading: false,
  hideValues: false,
  checkStateAsset: jest.fn()
};

describe("DetailsPositionCharts component", () => {
  const newProps = { ...props, loading: true };
  it("should match snapshot with name", () => {
    expect(render(<DetailsPositionCharts {...newProps} />)).toMatchSnapshot();
  });

  it("should match snapshot with name", () => {
    expect(render(<DetailsPositionCharts {...props} />)).toMatchSnapshot();
  });
});
