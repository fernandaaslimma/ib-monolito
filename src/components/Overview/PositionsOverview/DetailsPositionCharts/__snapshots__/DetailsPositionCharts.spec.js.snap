// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsPositionCharts component should match snapshot with name 1`] = `
Array [
  <div
    class="sc-bdVaJa NPQDb"
    data-test="undefined_loading_0"
  >
    <section
      class="sc-cSHVUG eykXTw"
    >
      <div
        class="sc-htpNat fcElnf"
      >
        <div
          class="sc-chPdSV eRYcUi"
          height="48"
          width="48"
        />
        <div
          class="sc-bxivhb izbgQu"
        >
          <div
            class="sc-ifAKCX dMjNLD"
          >
            <div
              class="sc-kAzzGY dNfLKP"
              height="14"
              width="80"
            />
          </div>
          <div
            class="sc-kAzzGY bnhxSk"
            height="12"
            width="80"
          />
        </div>
      </div>
    </section>
  </div>,
  <div
    class="sc-bdVaJa NPQDb"
    data-test="undefined_loading_1"
  >
    <section
      class="sc-cSHVUG eykXTw"
    >
      <div
        class="sc-htpNat fcElnf"
      >
        <div
          class="sc-chPdSV eRYcUi"
          height="48"
          width="48"
        />
        <div
          class="sc-bxivhb izbgQu"
        >
          <div
            class="sc-ifAKCX dMjNLD"
          >
            <div
              class="sc-kAzzGY dNfLKP"
              height="14"
              width="80"
            />
          </div>
          <div
            class="sc-kAzzGY bnhxSk"
            height="12"
            width="80"
          />
        </div>
      </div>
    </section>
  </div>,
  <div
    class="sc-bdVaJa NPQDb"
    data-test="undefined_loading_2"
  >
    <section
      class="sc-cSHVUG eykXTw"
    >
      <div
        class="sc-htpNat fcElnf"
      >
        <div
          class="sc-chPdSV eRYcUi"
          height="48"
          width="48"
        />
        <div
          class="sc-bxivhb izbgQu"
        >
          <div
            class="sc-ifAKCX dMjNLD"
          >
            <div
              class="sc-kAzzGY dNfLKP"
              height="14"
              width="80"
            />
          </div>
          <div
            class="sc-kAzzGY bnhxSk"
            height="12"
            width="80"
          />
        </div>
      </div>
    </section>
  </div>,
]
`;

exports[`DetailsPositionCharts component should match snapshot with name 2`] = `
Array [
  <div
    class="sc-bdVaJa NPQDb"
    data-test="undefined_0"
  >
    <section
      class="sc-cSHVUG eykXTw"
    >
      <div
        class="sc-htpNat fcElnf"
      >
        <div
          class="sc-iwsKbI jarkgW"
          color="#5b4a8a"
          data-test="chart_0"
          width="48"
        >
          <div
            class="sc-gZMcBi cPrKYz"
            width="48"
          >
            <p
              class="sc-gqjmRU dLFYij"
            >
              26.75%
            </p>
          </div>
        </div>
        <div
          class="sc-bxivhb izbgQu"
        >
          <div
            class="sc-ifAKCX dMjNLD"
            data-test="productName"
          >
            Conta Corrente
          </div>
          <div
            class="sc-kgoBCf flxnWk"
            data-test="HideableValue"
          >
            <span
              class="sc-dxgOiQ dNoZWu"
              value="631409.01"
            >
              R$
            </span>
            <span
              class="sc-kpOJdX czKNiG"
              data-test="value"
              value="631409.01"
            >
              631.409,01
            </span>
          </div>
        </div>
      </div>
    </section>
  </div>,
  <div
    class="sc-bdVaJa NPQDb"
    data-test="undefined_1"
  >
    <section
      class="sc-cSHVUG eykXTw"
    >
      <div
        class="sc-htpNat fcElnf"
      >
        <div
          class="sc-iwsKbI ksRqAj"
          color="#af5761"
          data-test="chart_1"
          width="48"
        >
          <div
            class="sc-gZMcBi cPrKYz"
            width="48"
          >
            <p
              class="sc-gqjmRU dLFYij"
            >
              --
            </p>
          </div>
        </div>
        <div
          class="sc-bxivhb izbgQu"
        >
          <div
            class="sc-ifAKCX dMjNLD"
            data-test="productName"
          >
            Conta Corrente
          </div>
          <div
            class="sc-kgoBCf flxnWk"
            data-test="HideableValue"
          >
            <span
              class="sc-dxgOiQ dNoZWu"
              value="631409.01"
            >
              R$
            </span>
            <span
              class="sc-kpOJdX czKNiG"
              data-test="value"
              value="631409.01"
            >
              631.409,01
            </span>
          </div>
        </div>
      </div>
    </section>
  </div>,
]
`;
